src/macros.rs

```diff
@@ -27,76 +27,57 @@ use crate::CStr8;
 /// ```
 #[macro_export]
 macro_rules! cstr8 {
-    ($s:expr) => {
-        unsafe {
-            const __CSTR: [$crate::u8; $crate::cstr_len_for($s)] = unsafe {
-                $crate::transmute_prefix($crate::Concat(
-                    $crate::str_as_byte_array::<{ $crate::str_len($s) }>($s),
-                    b'\0',
-                ))
-            };
-            $crate::CStr8::from_utf8_with_nul_unchecked(&__CSTR)
-        }
```

Comment:
I consider "macro `unsafe` leakage" as a kind of unsoundness, because it is a safe API that allows for client code that does not spell `unsafe` to invoke UB. In particular it will avoid triggering a `#![forbid(unsafe)]` declared for the crate.

The previous implementation being unsound was the initial impetus for my change. I also try to refactor `transmute` to something less general wherever I see it. 🚀

Reply:
FWIW, a re-test did show that the initial implementation wasn't actually unsound, as items do not inherit unsafety from separate enclosing items, and thus the length computation is done in a not-`unsafe` context. (So I didn't yank or file a RUSTSEC advisory.) It did lead me to re-discover that constant evaluation runs the function anyway even if it's a safety error, though, which is an interesting quirk. Not a harmful one, since consteval UB is limited to trashing the built binary (we won't get there because of the safety error) and cannot result in nasal-demons-at-a-distance, but certainly a potential surprise.

- I remembered correctly and I *had* went and audited this crate's safety hygiene in May 2023, resulting in filing this Rust issue about the behavior: https://github.com/rust-lang/rust/issues/111084

The new impl is certainly an improvement. I think the `transmute_prefix` union was required when I first wrote this code, but compiler improvements have since made that unnecessary (and I already wasn't claiming support for such a long-tail rust-version.)

Reply:
> as items do not inherit unsafety from separate enclosing items

Ah, correct, there is at least one place that doesn't have an `unsafe` context.

> https://github.com/rust-lang/rust/issues/111084

Good to know, and at least Rust CTFE defines behavior for all UB AFAIK. IIUC, the other two locations where `$s` was used would always evaluate the UB in a `const` context, even with this issue fixed.

> consteval UB is limited to trashing the built binary

Generally agreed especially given rustc's implementation. I'd be more concerned about other Rust compilers doing things like AOT compiling of `const fn` for CTFE and not catching UB.